File: backend/app/models/savings_analytics.py
from datetime import datetime
from pydantic import Field
from beanie import Indexed, Link
from app.models.base import BaseUUIDDocument
from app.models.cloud_resource import CloudResource
from app.models.base import get_utc_now

class SavingsAnalytics(BaseUUIDDocument):
    resource: Link[CloudResource]
    predicted_savings: float
    actual_savings: float
    percentage_saved: float
    calculation_date: Indexed(datetime) = Field(default_factory=get_utc_now)

    class Settings:
        name = "savings_analytics"
        indexes = [
            "calculation_date"
        ]

File: backend/src/models/SavingsAnalytics.ts
import mongoose, { Schema, Document } from 'mongoose';
import { BaseModelPlugin } from './BaseModelPlugin';
import crypto from 'crypto';
import { ICloudResource } from './CloudResource';

export interface ISavingsAnalytics extends Document<any> {
    _id: Buffer | string;
    resource: ICloudResource['_id'] | any;
    predicted_savings: number;
    actual_savings: number;
    percentage_saved: number;
    calculation_date: Date;
}

const savingsAnalyticsSchema = new Schema<ISavingsAnalytics>({
    _id: { type: 'UUID', default: () => crypto.randomUUID() },
    resource: { type: 'UUID', ref: 'CloudResource', required: true },
    predicted_savings: { type: Number, required: true },
    actual_savings: { type: Number, required: true },
    percentage_saved: { type: Number, required: true },
    calculation_date: { type: Date, default: Date.now, index: true }
}, { collection: 'savings_analytics' });

savingsAnalyticsSchema.plugin(BaseModelPlugin);

export const SavingsAnalytics = mongoose.model<ISavingsAnalytics>('SavingsAnalytics', savingsAnalyticsSchema);
